test/apps/iodemo/io_demo.cc

```diff
@@ -251,47 +309,92 @@ class P2pDemoCommon : public UcxContext {
         send_recv_data_as_chunks(conn, data_size, sn, XFER_TYPE_RECV, callback);
     }
 
-    uint32_t get_chunk_cnt(size_t data_size) {
-        return (data_size + opts().chunk_size - 1) / opts().chunk_size;
+    static uint32_t get_chunk_cnt(size_t data_size, size_t chunk_size) {
+        return (data_size + chunk_size - 1) / chunk_size;
     }
-    
+
     void send_data(UcxConnection* conn, size_t data_size, uint32_t sn,
                    UcxCallback* callback = EmptyCallback::get()) {
+        data_size = init_send_buffer(sn, data_size);
         send_data_as_chunks(conn, data_size, sn, callback);
     }
-    
+
     void recv_data(UcxConnection* conn, size_t data_size, uint32_t sn,
                    UcxCallback* callback = EmptyCallback::get()) {
+        if (opts().validate) {
+            data_size += sizeof(chk_hdr_t) + sizeof(tr_hdr_t);
+        }
         recv_data_as_chunks(conn, data_size, sn, callback);
     }
 
 protected:
-    const options_t          _test_opts;
-    MemoryPool<IoMessage>    _io_msg_pool;
+    static bool validate(uint32_t sn,      uint32_t peer_sn,
+                         uint16_t chk_sum, uint16_t peer_chk_sum) {
+        bool pass = true;
+
+        if (sn != peer_sn) {
+            pass = false;
+            LOG << "detected transaction mismatch " << sn << " != "
+                << peer_sn;
+        }
+
+        if (chk_sum != peer_chk_sum) {
+            pass = false;
+            LOG << "detected data corruption " << chk_sum << " != "
+                << peer_chk_sum;
+        }
+
+        return pass;
+    }
+
+private:
+    inline size_t init_send_buffer(uint32_t sn, size_t data_size) {
+        if (!opts().validate) {
+            return data_size;
+        }
+
+        chk_hdr_t *dhdr  = reinterpret_cast<chk_hdr_t*>(buffer());
+        tr_hdr_t *thdr   = reinterpret_cast<tr_hdr_t*>(dhdr + 1);
+        uint8_t *payload = reinterpret_cast<uint8_t*>(thdr + 1);
+
+        thdr->sn = sn;
+
+        for (size_t i = 0; i < data_size; ++i) {
+            payload[i] = IoDemoRandom::rand<uint8_t>();
+        }
+
+        dhdr->chk_sum = ucs_crc16(thdr, sizeof(*thdr) + data_size);
+        return data_size + sizeof(*dhdr) + sizeof(*thdr);
```

Comment:
can we make the headers part of data_size?
so when user specified size X that we'd send exactly size X - it's useful for selecting protocol thresholds

Reply:
done